cron/pubsub/subscriber.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2021 Security Scorecard Authors
+//
+// Licensed under the Apache License, Vershandlern 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permisshandlerns and
+// limitathandlerns under the License.
+
+package pubsub
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/ossf/scorecard/cron/data"
+	"gocloud.dev/pubsub"
+	_ "gocloud.dev/pubsub/gcppubsub" // Needed to link in GCP drivers.
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+var (
+	ErrorInReceive  = errors.New("error in Receive")
+	ErrorInParse    = errors.New("error during protojson.Unmarshal")
+	ErrorInShutdown = errors.New("error during shitdown")
+)
+
+type Subscriber interface {
+	SynchronousPull() (*data.ScorecardBatchRequest, error)
+	Ack()
+	Close() error
+}
+
+func CreateSubscriber(ctx context.Context, subscriptionURL string) (Subscriber, error) {
+	subscription, err := pubsub.OpenSubscription(ctx, subscriptionURL)
+	if err != nil {
+		return nil, fmt.Errorf("error during pubsub.OpenSubscription: %w", err)
+	}
+	ret := subscriberImpl{
+		ctx:          ctx,
+		subscription: subscription,
+	}
+	return &ret, nil
+}
+
+type receiver interface {
+	Receive(ctx context.Context) (*pubsub.Message, error)
+	Shutdown(ctx context.Context) error
+}
+
+type subscriberImpl struct {
+	ctx          context.Context
+	subscription receiver
+	msg          *pubsub.Message
+}
+
+func (subscriber *subscriberImpl) SynchronousPull() (*data.ScorecardBatchRequest, error) {
+	msg, err := subscriber.subscription.Receive(subscriber.ctx)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %v", ErrorInReceive, err)
+	}
+	subscriber.msg = msg
+
+	ret := &data.ScorecardBatchRequest{}
+	if err := protojson.Unmarshal(msg.Body, ret); err != nil {
+		return nil, fmt.Errorf("%w: %v", ErrorInParse, err)
+	}
+	return ret, nil
+}
+
+func (subscriber *subscriberImpl) Ack() {
+	subscriber.msg.Ack()
+}
+
+func (subscriber *subscriberImpl) Close() error {
+	err := subscriber.subscription.Shutdown(subscriber.ctx)
+	if err != nil {
+		return fmt.Errorf("%w: %v", ErrorInShutdown, err)
```

Comment:
Nit: this seems like a misuse of `Errorf`. The idea behind error wrapping is that you should be able to access a chain of underlying errors. This short circuits it to the one you've defined here (ErrorInShutdown).

To do what you want here, you probably instead want something like:

```go
type Error struct {
  Type string
  Err error
}

func (e *Error) Error() string{
  return fmt.Sprintf("%s: %v", e.Type, e.Err)
}

func (e *Error) Unwrap() error {
  return e.Err
}
```

Then construct/use them by doing:

```go
return &Error{type: "Shutdown", Err: error}

var pubsubErr pubsub.Error
if errors.As(err, &pubsubErr) && pubsubErr..... 
...
```

Reply:
I see. I was trying to test this in unit_test. I don't feel compelled complicating this code for the sake of a test. I will stick with the the simpler solution here.